src/snowflake/snowpark/_internal/packaging_utils.py

```diff
@@ -0,0 +1,467 @@
+                            for line in record_file.readlines()
+                        },
+                    )
+                    included_record_entries = []
+                    for record_entry in record_entries:
+                        record_entry_full_path = os.path.abspath(
+                            os.path.join(directory, record_entry),
+                        )
+                        # RECORD file might contain relative paths to items outside target folder. (ignore these)
+                        if (
+                            os.path.exists(record_entry_full_path)
+                            and directory in record_entry_full_path
+                            and len(record_entry) > 0
+                        ):
+                            included_record_entries.append(record_entry)
+                    package_req = Requirement.parse(package)
+                    return_dict[package_req] = included_record_entries
+    return return_dict
+
+
+def identify_supported_packages(
+    packages: List[Requirement],
+    valid_packages: Dict[str, List[str]],
+    native_packages: Set[str],
+) -> Tuple[List[Requirement], List[Requirement], List[Requirement]]:
+    supported_dependencies: List = []
+    dropped_dependencies: List = []
+    new_dependencies: List = []
+    for package in packages:
+        package_name = package.name
+        package_version_req = package.specs[0][1] if package.specs else None
+        if package_name in valid_packages:
+            if (package_version_req is None) or (
+                package_version_req in valid_packages[package_name]
+            ):
+                supported_dependencies.append(package)
+            elif package_name in native_packages:
+                # Native packages should anaconda dependencies if possible, even if the version is not available
+                _logger.warning(
+                    f"Package {package_name}(version {package_version_req}) is an unavailable native "
+                    f"dependency, switching to latest available version "
+                    f"{valid_packages[package_name][-1]} instead."
+                )
+                dropped_dependencies.append(package)
+                new_dependencies.append(Requirement.parse(package_name))
+                native_packages.remove(package_name)
+            elif package_name in DEFAULT_PACKAGES:
+                _logger.warning(
+                    f"Package {package_name}(version {package_version_req}) is an unavailable default "
+                    f"dependency, switching to latest available version "
+                    f"{valid_packages[package_name][-1]} instead."
+                )
+                dropped_dependencies.append(package)
+                new_dependencies.append(Requirement.parse(package_name))
+
+    return supported_dependencies, dropped_dependencies, new_dependencies
+
+
+# if package_name in valid_packages:
```

Comment:
Original code from SnowCLI:- any dependency that can be used via Anaconda channel is used despite versioning (I changed the code so that we now only use Anaconda if it's the right version or if it is a native dependency)